[bugfix] Fix Dissolve GDAL algorithm

agiudiceandrea opened the pull request:
## Description
Fixes [#19900](https://issues.qgis.org/issues/19900)

Statistics were erroneously computed on the dissolve field, rather than on the statistic attribute field as it should be.

Fixes [#19307](https://issues.qgis.org/issues/19307)

In order to properly handle filenames with white spaces (in the same way as Buffer, OffsetCurve, OneSideBuffer, PintsAlongLines algs).

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

agiudiceandrea committed:
Fix Dissolve GDAL algorithm

Fixes #19900

Statistics were erroneously computed on the dissolve field, rather than on the statistic attribute field as it should be.

agiudiceandrea committed:
Add test for GDAL Dissolve algorithm

nyalldawson commented on python/plugins/processing/tests/testdata/gdal_algorithm_tests.yaml:
```diff
@@ -733,3 +733,23 @@ tests:
         name: expected/gdal/points_along_lines.gml
         type: vector
+
+  - algorithm: gdal:dissolve
```
Instead of these tests, we've moved the gdal provider to use command line based tests (which are much easier to debug!). 

See e.g https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsTest.py#L1195

agiudiceandrea replied:
Is there a guide/manual about this new kind of test?
I've followed https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/README.md and http://qgis.it/processing_tests.html in order to write the tests.

nyalldawson replied:
Unfortunately not -- it's just a matter of copying the existing ones and adapting. I'm happy to write the initial tests for existing dissolve behavior if you need.

nyalldawson commented:
Fix looks good, thanks! It's just the test which needs updating.

agiudiceandrea committed:
Fix Dissolve GDAL alg: white spaces in filenames

Fixes #19307

Handle filenames with white spaces (in the same way as Buffer, OffsetCurve, OneSideBuffer, PintsAlongLines algs).

nyalldawson commented:
@agiudiceandrea  I've pulled these fixes into #8012 , along with new-style unit tests for all the dissolve options. That should give you a good indication of how these tests should be structured for gdal/ogr algorithms.

Thanks for your groundwork here!

agiudiceandrea commented:
Thank you @nyalldawson! I will follow your example on how to write the new-style tests.

In the meantime I've found new bugs about this and other gdal/ogr algorithms: https://issues.qgis.org/issues/19938
